sys/kerncore/src/lib.rs

```diff
@@ -0,0 +1,301 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+// Allow std-y things to be used in test. Note that this attribute is a bit of a
+// trap for the programmer, because rust-analyzer by default seems to build
+// things with test set. This means it's easy to introduce code incompatible
+// with no_std without your editor hassling you about it. Beware.
+#![cfg_attr(not(test), no_std)]
+
+pub trait UserSlice {
+    fn is_empty(&self) -> bool;
+    fn base_addr(&self) -> usize;
+    fn end_addr(&self) -> usize;
+}
+
+impl<T: UserSlice> UserSlice for &T {
+    fn is_empty(&self) -> bool {
+        (**self).is_empty()
+    }
+
+    fn base_addr(&self) -> usize {
+        (**self).base_addr()
+    }
+
+    fn end_addr(&self) -> usize {
+        (**self).end_addr()
+    }
+}
+
+pub trait MemoryRegion {
+    fn contains(&self, addr: usize) -> bool;
+    fn base_addr(&self) -> usize;
+    fn end_addr(&self) -> usize;
+}
+
+impl<T: MemoryRegion> MemoryRegion for &T {
+    fn contains(&self, addr: usize) -> bool {
+        (**self).contains(addr)
+    }
+
+    fn base_addr(&self) -> usize {
+        (**self).base_addr()
+    }
+
+    fn end_addr(&self) -> usize {
+        (**self).end_addr()
+    }
+}
+
+pub fn can_access<S, R>(
```

Comment:
might be nice to have a doc comment explaining what this does?

Reply:
Adding!